beacon_node/eth2-libp2p/src/rpc/protocol.rs

```diff
@@ -1,269 +1,318 @@
-    fn into(self) -> u64 {
-        self.0
+        Ok(ProtocolId {
+            message_name: protocol_list[4].into(),
+            version: protocol_list[5].into(),
+            encoding: protocol_list[6].into(),
+        })
     }
 }
 
-impl_encode_via_from!(RequestId, u64);
-impl_decode_via_from!(RequestId, u64);
-
-/// The RPC types which are sent/received in this protocol.
-#[derive(Debug, Clone)]
-pub enum RPCEvent {
-    Request {
-        id: RequestId,
-        method_id: u16,
-        body: RPCRequest,
-    },
-    Response {
-        id: RequestId,
-        method_id: u16, //TODO: Remove and process decoding upstream
-        result: RPCResponse,
-    },
+impl Into<RawProtocolId> for ProtocolId {
+    fn into(self) -> RawProtocolId {
+        format!(
+            "{}/{}/{}/{}",
+            PROTOCOL_PREFIX, self.message_name, self.version, self.encoding
+        )
+        .as_bytes()
+        .to_vec()
+    }
 }
 
-impl UpgradeInfo for RPCEvent {
-    type Info = &'static [u8];
-    type InfoIter = iter::Once<Self::Info>;
+/* Inbound upgrade */
 
-    #[inline]
-    fn protocol_info(&self) -> Self::InfoIter {
-        iter::once(b"/eth/serenity/rpc/1.0.0")
-    }
-}
+// The inbound protocol reads the request, decodes it and returns the stream to the protocol
+// handler to respond to once ready.
 
-type FnDecodeRPCEvent = fn(Vec<u8>, ()) -> Result<RPCEvent, DecodeError>;
+pub type InboundOutput<TSocket> = (RPCRequest, InboundFramed<TSocket>);
+pub type InboundFramed<TSocket> = Framed<upgrade::Negotiated<TSocket>, InboundCodec>;
+type FnAndThen<TSocket> = fn(
+    (Option<RPCRequest>, InboundFramed<TSocket>),
+) -> FutureResult<InboundOutput<TSocket>, RPCError>;
+type FnMapErr<TSocket> = fn(timeout::Error<(RPCError, InboundFramed<TSocket>)>) -> RPCError;
 
 impl<TSocket> InboundUpgrade<TSocket> for RPCProtocol
 where
     TSocket: AsyncRead + AsyncWrite,
 {
-    type Output = RPCEvent;
-    type Error = DecodeError;
-    type Future = upgrade::ReadOneThen<upgrade::Negotiated<TSocket>, (), FnDecodeRPCEvent>;
-
-    fn upgrade_inbound(self, socket: upgrade::Negotiated<TSocket>, _: Self::Info) -> Self::Future {
-        upgrade::read_one_then(socket, MAX_READ_SIZE, (), |packet, ()| Ok(decode(packet)?))
+    type Output = InboundOutput<TSocket>;
+    type Error = RPCError;
+
+    type Future = future::AndThen<
+        future::MapErr<
+            timeout::Timeout<stream::StreamFuture<InboundFramed<TSocket>>>,
+            FnMapErr<TSocket>,
+        >,
+        FutureResult<InboundOutput<TSocket>, RPCError>,
+        FnAndThen<TSocket>,
+    >;
+
+    fn upgrade_inbound(
+        self,
+        socket: upgrade::Negotiated<TSocket>,
+        protocol: RawProtocolId,
+    ) -> Self::Future {
+        // TODO: Verify this
+        let protocol_id =
+            ProtocolId::from_bytes(&protocol).expect("Can decode all supported protocols");
+
+        match protocol_id.encoding.as_str() {
+            "ssz" | _ => {
```

Comment:
Out of interest, why do you accept protocols that aren't `"ssz"`?

Reply:
The new spec, allows for arbitrary encodings. `ssz` is one... `ssz+snappy` is supposed to be another, but we may move compression down to the transport.